### PR TITLE
Change back SmallVec to Vec for JoinHashMap - Issue 5940

### LINKED ISSUE #5940: Change back SmallVec to Vec for JoinHashMap
### Is your feature request related to a problem or challenge?

_No response_

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #5940.

# Rationale for this change

After applying patches based on #5866 and #5904, based on the flame graph for q17 generated by 
`sudo CARGO_PROFILE_RELEASE_DEBUG=true cargo flamegraph --freq 500 --bin tpch -- benchmark datafusion --path ./data-parquet/ --format parquet --partitions 1 -q 17 --iterations 10`

flamegraph-5941-before

we found the bottleneck happens at **### physical_plan::joins::hash_join::update_hash**.

# What changes are included in this PR?

1. Change back SmallVec to Vec for JoinHashMap.

After applying the patch proposed from this PR, the flame graph becomes
flamegraph-5941-after

We can see the samples for the vector capacity changing significantly decreased. It's expected especially when doing joining with one big table and the other is relatively small.

To end to end query performance for q17 is improved from 7000ms to 6500ms.

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by yahoNanJing** [comment]
Hi @Dandandan and @alamb, could you help review this PR?

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 1]
```diff
@@ -543,7 +542,7 @@ async fn collect_left_input(
         )
     })? / 7)
         .next_power_of_two();
-    // 32 bytes per `(u64, SmallVec<[u64; 1]>)`
+    // 32 bytes per `(u64, Vec<[u64; 1]>)`
```
Comment (Dandandan) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:545]: This probably is a trade-off, joins with unique keys will probably get slower.

**Comment by alamb** [comment]
There is a CI check https://github.com/apache/arrow-datafusion/actions/runs/4657156069/jobs/8241558786?pr=5941 appears to be failing

Given that we have a benchmark that improves (q17) I think we should merge this PR, assuming we can resolve the CI failure satisfactorily. 

Thank you for this @yahoNanJing

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 2]
```diff
@@ -594,16 +593,17 @@ pub fn update_hash(
...
     // insert hashes to key of the hashmap
-    for (row, hash_value) in hash_values.iter().enumerate() {
-        let item = hash_map
-            .0
-            .get_mut(*hash_value, |(hash, _)| *hash_value == *hash);
+    let row_start = offset;
+    let row_end = offset + hash_values.len();
+    for (row, hash_value) in (row_start..row_end).zip(hash_values.iter()) {
+        // the hash value is the key, always true
+        let item = hash_map.0.get_mut(*hash_value, |_| true);
```
Comment (Dandandan) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:600]: The equality check might still be necessary.

**Comment by Dandandan** [comment]
It would be nice to show whether we are not getting regressions for other queries.

File: datafusion/core/src/physical_plan/joins/hash_join.rs (hunk 2 above)
Comment (yahoNanJing) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:600]: Thanks @Dandandan for pointing it out. You are right. It's my bad. The RawTable utilizes the open addressing policy for looking for matched entries. I'll revert this commit.

**Comment by Dandandan** [comment]
We have improved the datastructure of join in other ways already, so closing the PR in order to clean it up

**Pull request closed without merging**